mm/mm_heap/mm_sem.c

```diff
@@ -80,9 +80,21 @@ bool mm_takesemaphore(FAR struct mm_heap_s *heap)
 
   if (up_interrupt_context())
     {
-      /* Can't take semaphore in the interrupt handler */
+#if !defined(CONFIG_SMP) && defined(CONFIG_DEBUG_MM)
+      int val;
+
+      /* Check the semaphore value, if held by someone, then return false.
+       * Else, we can take it, return true.
+       */
+
+      _SEM_GETVALUE(&heap->mm_semaphore, &val);
```

Comment:
I just noticed that in https://github.com/apache/incubator-nuttx/pull/5504 we removed `return _SEM_TRYWAIT(&heap->mm_semaphore) >= 0;` line and replaced it with `_SEM_GETVALUE(&heap->mm_semaphore, &val);` + `return val > 0;`so the heap semaphore is not actually taken anymore in case of `else if (sched_idletask())` and `if (up_interrupt_context())`. Is this expected? Looks like it changes behavior of the system tremendously

Reply:
Should we get back `return _SEM_TRYWAIT(&heap->mm_semaphore) >= 0;` at least for the `else if (sched_idletask())`?

Reply:
There is a special case for `mm_free()`:
```
  if (mm_takesemaphore(heap) == false)
    {
      kasan_unpoison(mem, mm_malloc_size(mem));

      /* We are in IDLE task & can't get sem, or meet -ESRCH return,
       * which means we are in situations during context switching(See
       * mm_takesemaphore() & getpid()). Then add to the delay list.
       */

      mm_add_delaylist(heap, mem);
      return;
    }
```
I think we should try to get back `return _SEM_TRYWAIT(&heap->mm_semaphore) >= 0;` in case of `else if (sched_idletask())`

Reply:
Yes, that's the change, but in @masayuki2009 's case, none will take mm_sem from IRQ context.
And also, mm_takesemaphore will return false, when meet SMP mode.

```
 76 bool mm_takesemaphore(FAR struct mm_heap_s *heap)
 77 {
 78 #if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
 79   /* Check current environment */
 80 
 81   if (up_interrupt_context())
 82     {
 83 #if !defined(CONFIG_SMP) && defined(CONFIG_DEBUG_MM)
 84       int val;
 85 
 86       /* Check the semaphore value, if held by someone, then return false.
 87        * Else, we can take it, return true.
 88        */
 89 
 90       _SEM_GETVALUE(&heap->mm_semaphore, &val);
 91 
 92       return val > 0;
 93 #else
 94       /* Can't take semaphore in SMP interrupt handler */
 95 
 96       return false;
 97 #endif
 98     }
```

Reply:
Maybe we have interrupt nesting enabled in faulty case?